### ISSUE TITLE
attrib() interpolation with curveVertex shapes

### ISSUE THREAD

**Issue by AmnonOwed** [processing/processing#5091]
I've made a small code example that shows the issue, see below. When using custom attrib() with vertex() the interpolation works as expected (see the shape on the left). However when using the exact same values with curveVertex() things take a turn for the messy (see the shape on the right).

My end goal is to have  a smooth shape (using curveVertex, as seen on the right) with a smooth gradient (using attrib, as seen on the left). While I've managed both, unfortunately they don't seem to work at the same time.

@codeanticode Thoughts/suggestions?

![attrib](https://cloud.githubusercontent.com/assets/4075846/26521395/f1f1ee6a-42e7-11e7-8319-cac1f3953946.jpg)

## Sketch
``` processing
PShader shader;

float[][] pos;
float[][] uv;

void setup() {
  size(1200, 600, P2D);
  smooth(8);
  shader = loadShader("frag.glsl", "vert.glsl");
  shader(shader);
  noStroke();
}

void draw() {
  int numPoints = int(map(mouseX, 0, width, 3, 30));
  calculatePositionsAndUVs(numPoints);
  background(255);

  // vertex shape
  beginShape();
  for (int i=0; i<pos.length; i++) {
    float x = pos[i][0];
    float y = pos[i][1];
    float uvx = uv[i][0];
    float uvy = uv[i][1];
    attrib("uv", uvx, uvy);
    vertex(x, y);
  }
  endShape();

  // curve vertex shape (using exact same positions and UVs)
  translate(width/2, 0);
  beginShape();
  for (int i=0; i<pos.length; i++) {
    float x = pos[i][0];
    float y = pos[i][1];
    float uvx = uv[i][0];
    float uvy = uv[i][1];
    attrib("uv", uvx, uvy);
    curveVertex(x, y);
  }
  for (int i=0; i<3; i++) {
    float x = pos[i][0];
    float y = pos[i][1];
    float uvx = uv[i][0];
    float uvy = uv[i][1];
    attrib("uv", uvx, uvy);
    curveVertex(x, y);
  }
  endShape();
}

void calculatePositionsAndUVs(int numPoints) {
  pos = new float[numPoints][2];
  uv = new float[numPoints][2];
  float radius = height/2 - 50;
  float theta = TWO_PI / numPoints;
  for (int i=0; i<numPoints; i++) {
    float x = height/2 + sin(i * theta) * radius;
    float y = height/2 + cos(i * theta) * radius;
    pos[i][0] = x;
    pos[i][1] = y;
    float uvx = map(x, 0, height, 0, 1);
    float uvy = map(y, 0, height, 0, 1);
    uv[i][0] = uvx;
    uv[i][1] = uvy;
  }
}
```

## Vertex Shader
``` glsl
uniform mat4 transformMatrix;

attribute vec4 position;
attribute vec2 uv;

varying vec2 vertUV;

void main() {
  gl_Position = transformMatrix * position;
  vertUV = uv;
}
```

## Fragment Shader
``` glsl
varying vec2 vertUV;

uniform vec3 color0 = vec3(1.0, 0.0, 0.0);
uniform vec3 color1 = vec3(0.0, 1.0, 0.0);
uniform vec3 color2 = vec3(0.0, 0.0, 1.0);
uniform vec3 color3 = vec3(1.0, 1.0, 0.0);

void main(void) {
	vec2 coordinate = vertUV;
	
	vec3 colorMix_01 = mix(color0, color1, coordinate.x);
	vec3 colorMix_23 = mix(color2, color3, coordinate.x);
	vec3 finalColor = mix(colorMix_01, colorMix_23, coordinate.y);

	gl_FragColor = vec4(finalColor, 1.0);
}
```

**Comment by AmnonOwed** [comment]
My current workaround is to drop the use of attrib() completely and use gl_Position instead, since it's interpolated correctly, even for curveVertex() shapes. As a result I've moved most of the calculations into the shader. In this case it works as a workaround, because the gradient was already position-based anyway.

![attrib2](https://cloud.githubusercontent.com/assets/4075846/26523378/b1f94e78-4316-11e7-84fb-ff1a4393ee9e.jpg)

**Comment by codeanticode** [comment]
@AmnonOwed The problem is that each curveVertex() call generates a [series of vertices](https://github.com/processing/processing/blob/master/core/src/processing/opengl/PGraphicsOpenGL.java#L12757) in the tessellated geometry, along the path of the Catmull-Rom spline, and all these vertices share the same custom attributes. I think we would need to interpolate the attributes according to the spline parameter.

**Comment by AmnonOwed** [comment]
Ah, right, that must be it. So currently there is no interpolation at all for attributes.

I'm wondering what is a logical interpolation method for the attributes on splines. See the image below.

![attribinterpolation](https://cloud.githubusercontent.com/assets/4075846/26719736/01b1a834-4786-11e7-950d-ba1e6adf2139.jpg)
Take point K. If you use the relative position on the path segment from B to C (ranging from 0 to 1) as the interpolation parameter for all attrib values, isn't there a risk that you end up with interpolated attrib values that are actually for point T? Is this solution what you mean by spline parameter? It's an improvement over no interpolation, but still doesn't seem like the correct result.

If you instead use the xyz-coordinates of K and remap it between 0 and 1 using the appropiate coordinates of B and C respectively and then use that value as the interpolation parameter to calculate the attrib value, then you might get results that are actually for point K. But there is an implicit assumption behind this solution that the first attib is related to x, the second to y and the third to z. In my use-case that is definitely the case. Generally speaking it may also be the most 'logical' way of interpolation, because it takes into account the final position of the point on the spline, which feels correct. However I don't know if it would make sense for every use-case. Also, what if there is a discrepancy between the number of attributes and the number of vertex dimensions? And finally, suppose you use this solution on point L, then your second attribute value will be negative (because all y-positions of the spline are negative compared to the vertices A and B).

So there are at least three interpolation solutions:
1. No interpolation
2. Interpolation based on the relative position in the segment
3. Interpolation based on xyz-coordinate (remapping)

Each solution might be appropiate for certain use-cases. I'm inclined to choose the third solution, because it feels the most correct. However my perspective may be influenced by my use-case and there are some issues with this solution as well.

@codeanticode Thoughts?

**Comment by codeanticode** [comment]
@AmnonOwed yes, this is probably the reason why I did not implement attribute interpolation, since it is not uniquely defined... 

Now, what about using the chord length as the interpolation value, instead of the spline parameter? Of course, it should be normalized by the total chord length between the endpoint vertices for each curve segment. Wouldn't this be a more "natural" default that would not depend on the number of coordinates?

**Comment by AmnonOwed** [comment]
Yes, I suppose that is a suitable default solution, especially because it's independent from the number of attributes and coordinates. And it's an improvement over the current solution of no interpolation, because transitions between attributes will become much smoother.

A downside of this solution is that it doesn't really take into account the actual coordinates of the new vertex it's calculating the attrib value(s) for. So for some people it may give a different result than expected / desired. However one could say that for those use-cases the workaround I proposed above (use gl_Position instead of attrib() and do the calculations in the shader) is the preferred route. Perhaps this could be clarified in the attrib() reference eventually.

**Comment by JakubValtar** [comment]
@codeanticode:
> Now, what about using the (non-linear) chord length as the interpolation parameter, instead of the (linear) spline parameter? Of course, it should be normalized by the total chord length between the endpoint vertices for each curve segment. Wouldn't this be a more "natural" default that will not depend on the number of coordinates?

I think this would be the most reasonable solution. We calculate all the points anyway so calculating their distances and interpolating the attribute should not add any unreasonable overhead.

@AmnonOwed:
> A downside of this solution is that it doesn't really take into account the actual coordinates of the new vertex it's calculating the attrib value(s) for. So for some people it may give a different result than expected / desired.

Since it is not uniquely defined there will for sure be people who would get something else than what they expected.

We have to assign a value between 0 and 1 to every curve point and then just mix the attrib values of the two endpoints. We can't go out of the 0-1 range because then we would start extrapolating and the values might not be defined (e.g. colors).

I was thinking about projecting the point position on a line between the two endpoints and taking the parameter from there, but it would fail for exactly this reason, because the curve can go beyond the endpoint and then bend back.

@AmnonOwed:
> However one could say that for those use-cases the workaround I proposed above (use gl_Position instead of attrib() and do the calculations in the shader) is the preferred route.

Another option would be to calculate the point positions yourself, assign attributes you want, and draw it with `vertex()` only. Although it's gonna make the code a bit more complicated, we provide both `curvePoint()` and `bezierPoint()` so it should not be that hard.

**Comment by AmnonOwed** [comment]
All right, sounds good.

Ah, good point. Indeed, that's another way to do it with the advantage of a little more direct control inside the sketch itself.
